DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc

```diff
@@ -0,0 +1,122 @@
+//
+#include "DataFormats/Math/interface/approx_atan2.h"
+#include "CUDADataFormats/Common/interface/Product.h"
+#include "CUDADataFormats/TrackingRecHit/interface/TrackingRecHit2DHeterogeneous.h"
+// Geometry
+#include "DataFormats/TrackerCommon/interface/TrackerTopology.h"
+#include "DataFormats/SiPixelDetId/interface/PixelSubdetector.h"
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+#include "Geometry/CommonDetUnit/interface/PixelGeomDetUnit.h"
+#include "Geometry/CommonTopologies/interface/PixelTopology.h"
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "SiPixelPhase1MonitorRecHitsSoABase.h"
+
+SiPixelPhase1MonitorRecHitsSoABase::SiPixelPhase1MonitorRecHitsSoABase(const edm::ParameterSet& iConfig)
+    : geomToken_(esConsumes<TrackerGeometry, TrackerDigiGeometryRecord, edm::Transition::BeginRun>()),
+      topoToken_(esConsumes<TrackerTopology, TrackerTopologyRcd, edm::Transition::BeginRun>()) {
+  topFolderName_ = iConfig.getParameter<std::string>("TopFolderName");  //"SiPixelHeterogeneous/PixelRecHitsSoA";
+  onGPU_ = iConfig.getParameter<bool>("onGPU");                         //unused now, but may be needed
+}
+
+//
+// destructor
+//
+SiPixelPhase1MonitorRecHitsSoABase::~SiPixelPhase1MonitorRecHitsSoABase() {
+  // do anything here that needs to be done at desctruction time
+  // (e.g. close files, deallocate resources etc.)
+  edm::LogInfo("SiPixelPhase1MonitorRecHitsSoABase") << ">>> Destroy SiPixelPhase1MonitorRecHitsSoABase ";
+}
+void SiPixelPhase1MonitorRecHitsSoABase::dqmBeginRun(const edm::Run& iRun, const edm::EventSetup& iSetup) {
+  tkGeom_ = &iSetup.getData(geomToken_);
+  tTopo_ = &iSetup.getData(topoToken_);
+}
+//
+// -- Book Histograms
+//
+void SiPixelPhase1MonitorRecHitsSoABase::bookHistograms(DQMStore::IBooker& iBook,
+                                                        edm::Run const& iRun,
+                                                        edm::EventSetup const& iSetup) {
+  iBook.cd();
+  iBook.setCurrentFolder(topFolderName_);
+
+  // clang-format off
+  //Global
+  hnHits = iBook.book1D("nHits", "RecHits per event;RecHits;#events", 200, 0, 5000);
+  hBFposZP = iBook.book2D("recHitsGlobalPosZP", "RecHits position Global;Z;#phi", 1000, -50, 50, 200,-3.2,3.2);
+  hBFposZR = iBook.book2D("recHitsGlobalPosZR", "RecHits position Global;Z;R", 1000, -50, 50, 200,-20,20);
+  //Barrel
+  hBposXY = iBook.book2D("recHitsBarrelPosXY", "RecHits position Barrel;X;Y", 200, -20, 20, 200,-20,20);
+  hBposZP = iBook.book2D("recHitsBarrelPosZP", "RecHits position Barrel;Z;#phi", 300, -30, 30, 200,-3.2,3.2);
+  hBcharge = iBook.book1D("recHitsBarrelCharge", "RecHits Charge Barrel;Charge;#events", 250, 0, 100000);
+  hBsizex = iBook.book1D("recHitsBarrelSizex", "RecHits SizeX Barrel;SizeX;#events", 50, 0, 50);
+  hBsizey = iBook.book1D("recHitsBarrelSizey", "RecHits SizeY Barrel;SizeY;#events", 50, 0, 50);
+  //Barrel Layer
+  for(int il=0;il<4;il++){
```

Comment:
this you can take from the ES.
See e.g. at:

https://github.com/cms-sw/cmssw/blob/772910aeac89026866ae4c7a7246e33f49a8e300/CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLHarvester.cc#L139-L146